build/README.md

```diff
@@ -0,0 +1,171 @@
+        "fsdp_state_dict_type": "FULL_STATE_DICT",
+        "fsdp_cpu_ram_efficient_loading": true,
+        "fsdp_sync_module_states": true
+    },
+    "multi_gpu": true,
+    "model_name_or_path": "/llama/13B",
+    "training_data_path": "/data/twitter_complaints.json",
+    "output_dir": "/output/llama-7b-pt-multigpu",
+    "num_train_epochs": 5.0,
+    "per_device_train_batch_size": 4,
+    "per_device_eval_batch_size": 4,
+    "gradient_accumulation_steps": 4,
+    "save_strategy": "epoch",
+    "learning_rate": 0.03,
+    "weight_decay": 0.0,
+    "lr_scheduler_type": "cosine",
+    "logging_steps": 1.0,
+    "packing": false,
+    "include_tokens_per_second": true,
+    "response_template": "\n### Label:",
+    "dataset_text_field": "output",
+    "use_flash_attn": true,
+    "torch_dtype": "bfloat16",
+    "tokenizer_name_or_path": "/llama/13B"
+}
+```
+
+When `multi_gpu` is set to true, the [FSDP config](https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/fixtures/accelerate_fsdp_defaults.yaml) is used by default. Any of these values can be overwritten by passing in flags via the JSON config or by passing in your own config file using key `config_file`.
+
+If `multi_gpu` is set and `num_processes` is not explicitly set, the number of processes/GPUs will be determined by the number of GPUs available via `torch.cuda.device_count()`.
+
+If `multi_gpu` is not set, the script will assume single-GPU and run with `num_processes=1`.
+
+Note that `num_processes` which is the total number of processes to be launched in parallel, should match the number of GPUs to run on. The number of GPUs used can also be set by setting environment variable `CUDA_VISIBLE_DEVICES`.
+
+
+## Building the Image
+
+With docker, build the image at the top level with:
+
+```sh
+docker build . -t sft-trainer:mytag -f build/Dockerfile
+```
+
+## Running the Image
+
+Run sft-trainer-image with the JSON env var and mounts set up.
+
+```sh
+docker run -v config.json:/app/config.json -v $MODEL_PATH:/model -v $TRAINING_DATA_PATH:/data/twitter_complaints.json --env SFT_TRAINER_CONFIG_JSON_PATH=/app/config.json sft-trainer:mytag
+```
+
+This will run `accelerate_launch.py` with the JSON config passed.
+
+An example Kubernetes Pod for deploying sft-trainer which requires creating PVCs with the model and input dataset and any mounts needed for the outputted tuned model:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+name: sft-trainer-config
+data:
+config.json: |
+    {
+        "accelerate_launch_args": {
+            "num_machines": 1,
+            "main_process_port": 1234,
+            "num_processes": 2,
+            "use_fsdp": true,
+            "fsdp_backward_prefetch_policy": "TRANSFORMER_BASED_WRAP",
+            "fsdp_sharding_strategy": 1,
+            "fsdp_state_dict_type": "FULL_STATE_DICT",
+            "fsdp_cpu_ram_efficient_loading": true,
+            "fsdp_sync_module_states": true
+        },
+        "multi_gpu": true,
```

Comment:
I'm okay with either, whichever is more easily understandable. Distributed technically is more accurate

Reply:
distributed can mean both multi-gpu or multi-mode. If we dont allow user to set `num_machines`, then I think the `accelerate` accurate term is `multi_gpu`

Reply:
Note that I am not passing `--multi_gpi` in as a flag to `accelerate launch`. I know this flag exists but I saw that passing in the flag gave poor behavior and I believe it's because the `distributed_type` gets set to `MULTI_GPU` instead of `FSDP` so overriding the config_file value. And I believe MULTI_GPU uses DPP instead of FSDP so I was hitting into memory issues when adding flag `--mutli_gpu`

Reply:
@anhuong yes you are right, if `args.mult_gpu == True`, then `args.use_fsdp` will not be set, see [here](https://github.com/huggingface/accelerate/blob/b634388ef1e5d3631e88fdc08eda321c1c69e874/src/accelerate/commands/launch.py#L900).

Reply:
I think lets stay with `multi_gpu` (instead of distributed), thanks for the discussion on this!

Reply:
also important to note that I get num_processes warning for multiGPU when single GPU set. I passed in `num_processes=1` into a pod that had 2 GPUs available and got below warning message
```
   WARNING:accelerate.commands.launch:The following values were not passed to `accelerate launch` and had defaults used instead:
		More than one GPU was found, enabling multi-GPU training.
		If this was unintended please pass in `--num_processes=1`.
	`--num_machines` was set to a value of `1`
	`--mixed_precision` was set to a value of `'no'`
	`--dynamo_backend` was set to a value of `'no'`
```
This means that internally multi_gpu is being set, however the value of num_processes is not being overridden otherwise we would see message `--num_machines was set to a value of 2`. However, I think this means DPP is being used. This is an important usage note on if multi-gpu is being used and is on the user to set resources appropriately. Thoughts on this behavior?